server/src/test/java/org/elasticsearch/action/admin/ReloadSecureSettingsIT.java

```diff
@@ -0,0 +1,426 @@
+                                assertThat(nodeResponse.reloadException(), nullValue());
+                            }
+                        } catch (final AssertionError e) {
+                            reloadSettingsError.set(e);
+                        } finally {
+                            latch.countDown();
+                        }
+                    }
+
+                    @Override
+                    public void onFailure(Exception e) {
+                        reloadSettingsError.set(new AssertionError("Nodes request failed", e));
+                        latch.countDown();
+                    }
+                });
+        latch.await();
+        if (reloadSettingsError.get() != null) {
+            throw reloadSettingsError.get();
+        }
+    }
+
+    private SecureSettings loadResourceAsKeystore(Environment environment, String resourceName)
+            throws IOException, GeneralSecurityException {
+        try (InputStream keystore = ReloadSecureSettingsIT.class.getResourceAsStream(resourceName)) {
+            if (Files.exists(environment.configFile()) == false) {
+                Files.createDirectory(environment.configFile());
+            }
+            Files.copy(keystore, KeyStoreWrapper.keystorePath(environment.configFile()), StandardCopyOption.REPLACE_EXISTING);
+        }
+        // read the keystore from the testcase code (not node code)
+        final KeyStoreWrapper keyStoreWrapper = KeyStoreWrapper.load(environment.configFile());
+        keyStoreWrapper.decrypt(new char[0]);
+        return keyStoreWrapper;
+    }
+
+    public static class CountingReloadablePlugin extends Plugin implements ReloadablePlugin {
+
+        private volatile int reloadCount;
+
+        public CountingReloadablePlugin() {
+        }
+
+        @Override
+        public void reload(Settings settings) throws Exception {
+            reloadCount++;
+        }
+
+        public int getReloadCount() {
+            return reloadCount;
+        }
+
+    }
+
+    public static class MockReloadablePlugin extends CountingReloadablePlugin {
+
+        static final Setting<SecureString> DUMMY_SECRET_SETTING = SecureSetting.secureString("dummy_secret_setting", null);
+        private volatile String dummySecretValue;
+
+        public MockReloadablePlugin() {
+        }
+
+        @Override
+        public void reload(Settings settings) throws Exception {
+            super.reload(settings);
+            dummySecretValue = DUMMY_SECRET_SETTING.get(settings).toString();
+        }
+
+        @Override
+        public List<Setting<?>> getSettings() {
+            return Arrays.asList(DUMMY_SECRET_SETTING);
+        }
+
+        public String getDummySecretValue() {
+            return dummySecretValue;
+        }
+
+    }
+
+    public static class MisbehavingReloadablePlugin extends CountingReloadablePlugin {
+
+        private volatile boolean sulky = false;
```

Comment:
can we rename this to `shouldThrow`? Then just have a setter for the method? Also, let's make this non-volatile and the methods synchronized since we technically do a CAS in reload

Reply:
Yes, I've been lax here.